batch/batch/front_end/front_end.py

```diff
@@ -1420,75 +1765,107 @@ async def create_update(request, userdata):
+
+    update_id, start_job_group_id, start_job_id = await _create_batch_update(
+        batch_id, update_spec['token'], n_jobs, n_job_groups, user, db
+    )
+    return json_response({
+        'update_id': update_id,
+        'start_job_group_id': start_job_group_id,
+        'start_job_id': start_job_id,
+    })
 
 
 async def _create_batch_update(
-    batch_id: int, update_token: str, n_jobs: int, user: str, db: Database
-) -> Tuple[int, int]:
+    batch_id: int, update_token: str, n_jobs: int, n_job_groups: int, user: str, db: Database
+) -> Tuple[int, int, int]:
     @transaction(db)
     async def update(tx: Transaction):
-        assert n_jobs > 0
+        assert n_jobs > 0 or n_job_groups > 0
         record = await tx.execute_and_fetchone(
             """
-SELECT update_id, start_job_id FROM batch_updates
+SELECT update_id, start_job_id, start_job_group_id FROM batch_updates
 WHERE batch_id = %s AND token = %s;
 """,
             (batch_id, update_token),
         )
 
         if record:
-            return record['update_id'], record['start_job_id']
+            return (record['update_id'], record['start_job_id'], record['start_job_group_id'])
 
         # We use FOR UPDATE so that we serialize batch update insertions
         # This is necessary to reserve job id ranges.
         # We don't allow updates to batches that have been cancelled
         # but do allow updates to batches with jobs that have been cancelled.
         record = await tx.execute_and_fetchone(
             """
-SELECT job_groups_cancelled.id IS NOT NULL AS cancelled
+SELECT cancelled_t.cancelled IS NOT NULL AS cancelled
 FROM batches
-LEFT JOIN job_groups_cancelled ON batches.id = job_groups_cancelled.id
-WHERE batches.id = %s AND user = %s AND NOT deleted
+LEFT JOIN LATERAL (
+  SELECT 1 AS cancelled
+  FROM job_group_self_and_ancestors
+  INNER JOIN job_groups_cancelled
+    ON job_group_self_and_ancestors.batch_id = job_groups_cancelled.id AND
+      job_group_self_and_ancestors.ancestor_id = job_groups_cancelled.job_group_id
+  WHERE batches.id = job_group_self_and_ancestors.batch_id AND
+    job_group_self_and_ancestors.job_group_id = %s
+) AS cancelled_t ON TRUE
+WHERE batches.id = %s AND batches.user = %s AND NOT deleted
 FOR UPDATE;
 """,
-            (batch_id, user),
+            (ROOT_JOB_GROUP_ID, batch_id, user),
         )
         if not record:
             raise web.HTTPNotFound()
         if record['cancelled']:
-            raise web.HTTPBadRequest(reason='Cannot submit new jobs to a cancelled batch')
+            raise web.HTTPBadRequest(reason='Cannot submit new jobs or job groups to a cancelled batch')
 
         now = time_msecs()
 
         record = await tx.execute_and_fetchone(
             """
-SELECT update_id, start_job_id, n_jobs FROM batch_updates
+SELECT update_id, start_job_id, n_jobs, start_job_group_id, n_job_groups
+FROM batch_updates
 WHERE batch_id = %s
 ORDER BY update_id DESC
-LIMIT 1;
+LIMIT 1
+FOR UPDATE;
```

Comment:
No. I shouldn't have added this lock. The comment in the block above explains it.

```
        # We use FOR UPDATE so that we serialize batch update insertions
        # This is necessary to reserve job id and job group id ranges.
        # We don't allow updates to batches that have been cancelled
        # but do allow updates to batches with jobs that have been cancelled.
```